eventauth/eventauth.go

```diff
@@ -362,3 +368,199 @@ func (e *eventAllower) commonChecks(event Event) error {
 
 	return nil
 }
+
+// A membershipAllower has the information needed to authenticate a m.room.member event
+type membershipAllower struct {
+	// The user ID of the user who's membership is changing.
+	targetID string
+	// The user ID of the user who sent the membership event.
+	senderID string
+	// The membership of the user who sent the membership event.
+	senderMember memberContent
+	// The previous membership of the user who's membership is changing.
+	oldMember memberContent
+	// The new membership of the user if this event is accepted.
+	newMember memberContent
+	// The m.room.create content for the room.
+	create createContent
+	// The m.room.power_levels content for the room.
+	powerLevels powerLevelContent
+	// The m.room.join_rules content for the room.
+	joinRule joinRuleContent
+}
+
+// newMembershipAllower loads the information needed to authenticate the m.room.member event
+// from the auth events.
+func newMembershipAllower(authEvents AuthEvents, event Event) (m membershipAllower, err error) {
+	if event.StateKey == nil {
+		err = errorf("m.room.member must be a state event")
+		return
+	}
+	m.targetID = *event.StateKey
```

Comment:
Yeah, Maybe we should start a tracking these omissions somewhere, I'd quite like to update this and synapse's checks in parallel to keep them in sync.